src/pykeen/training/callbacks.py

```diff
@@ -374,9 +384,165 @@ def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # n
+
+    # docstr-coverage: inherited
+    def post_batch(self, epoch: int, batch, **kwargs: Any) -> None:  # noqa: D102
+        # when called by batch_size_search(), the parameter update should not be applied.
+        if self.step:
+            # pre-step callbacks
+            for cb in self.pre_step_callbacks:
+                cb.pre_step(epoch=epoch, **kwargs)
+            # update parameters according to optimizer
+            self.optimizer.step()
+
+        # After changing applying the gradients to the embeddings, the model is notified that the forward
+        # constraints are no longer applied
+        # TODO: do we want that when self.step is not set?
+        self.model.post_parameter_update()
+
+
+class LearningRateSchedulerTrainingCallback(TrainingCallback):
+    """Update learning rate scheduler."""
+
+    # docstr-coverage: inherited
+    def post_epoch(self, epoch: int, epoch_loss: float, **kwargs: Any) -> None:  # noqa: D102
+        if self.training_loop.lr_scheduler is not None:
+            self.training_loop.lr_scheduler.step(epoch=epoch)
+
+
+def _hasher(kwargs: Mapping[str, Any]) -> int:
+    # do not share optimal parameters across different training loops
+    return id(kwargs["training_loop"])
+
+
+@maximize_memory_utilization(parameter_name=("batch_size", "slice_size"), hasher=_hasher)
+@torch.inference_mode()
+def _validation_loss_amo_wrapper(
+    training_loop: "pykeen.training.TrainingLoop",
+    triples_factory: CoreTriplesFactory,
+    batch_size: int,
+    slice_size: int,
+    label_smoothing: float,
+    **kwargs,
+) -> float:
+    """Calculate validation loss with automatic batch size optimization."""
+    return training_loop._train_epoch(
+        # todo: create dataset only once
+        batches=training_loop._create_training_data_loader(
+            triples_factory=triples_factory, batch_size=batch_size, drop_last=False, **kwargs
+        ),
+        label_smoothing=label_smoothing,
+        # no callbacks, no epoch
+        callbacks=MultiTrainingCallback(),
+        epoch=-1,
+        # no sub-batching (for evaluation, we can just reduce batch size without any effect)
+        sub_batch_size=None,
+        slice_size=slice_size if training_loop.supports_slicing else None,
+        # this is handled by the AMO wrapper
+        only_size_probing=False,
+        # no backward passes
+        backward=False,
+    )
+
+
+class ValidationLossTrainingCallback(TrainingCallback):
+    """
+    Calculate loss on a development set.
+
+    .. code-block ::
+
+        from pykeen.datasets import get_dataset
+        from pykeen.pipeline import pipeline
+
+        dataset = get_dataset(dataset="nations")
+        pipeline(
+            dataset=dataset,
+            model="mure",
+            training_kwargs=dict(
+                callbacks="validation-loss",
+                callback_kwargs=dict(triples_factory=dataset.validation),
+            ),
+            result_tracker="console",
+        )
```

Comment:
can you please explain how to get the validation losses after the fact as a list

Reply:
I guess someone might 1 use this in combination with a result tracker or 2 want to make their own charts or something

Reply:
how about adding that to `notebooks`? (using the `# %%` style for better VCS): 49ae07e

Reply:
![myplot](https://github.com/pykeen/pykeen/assets/36984984/a6e63e5c-f769-4d7c-a2d5-35c380bea053)

Reply:
This is great, but I think it's ideal to keep the code examples in with the code itself so when people are looking through the docs they see it. It's also okay if we double it

Reply:
Hm, the notebook is now 45 lines (tbf with formatting), and it is really messy to get the code well-formatted into rst...

Reply:
The short version without the post-processing / plotting logic is already inside the docstring